src/algorithms/ctmrg.jl

```diff
@@ -1,321 +1,265 @@
+    ϵold = 1.0
+
+    env = deepcopy(envinit)
+    for i in 1:(alg.maxiter)
+        env, iterinfo = ctmrg_iter(state, env, alg)  # Grow and renormalize in all 4 directions
+
+        # Compute convergence criteria and take max (TODO: How should we handle logging all of this?)
+        Δϵ = abs((ϵold - iterinfo.ϵ) / ϵold)
+        normnew = contract_ctmrg(state, env)
+        Δnorm = abs(normold - normnew)
+        CSnew = tsvd(env.corners[NORTHWEST]; alg=TensorKit.SVD())[2]
+        ΔCS = norm(CSnew - CSold)
+        TSnew = tsvd(env.edges[NORTH]; alg=TensorKit.SVD())[2]
+        ΔTS = norm(TSnew - TSold)
+        (max(Δnorm, ΔCS, ΔTS) < alg.tol && i > alg.miniter) && break  # Converge if maximal Δ falls below tolerance
+
+        # Print verbose info
+        ignore_derivatives() do
+            alg.verbose > 1 && @printf(
+                "CTMRG iter: %3d   norm: %.2e   Δnorm: %.2e   ΔCS: %.2e   ΔTS: %.2e   ϵ: %.2e   Δϵ: %.2e\n",
+                i,
+                abs(normnew),
+                Δnorm,
+                ΔCS,
+                ΔTS,
+                iterinfo.ϵ,
+                Δϵ
+            )
+            alg.verbose > 0 &&
+                i == alg.maxiter &&
+                @warn(
+                    "CTMRG reached maximal number of iterations at (Δnorm=$Δnorm, ΔCS=$ΔCS, ΔTS=$ΔTS)"
+                )
         end
 
-        new_norm = contract_ctrmg(envs)
-
-        err = abs(old_norm - new_norm)
-        dϵ = abs((ϵ₁ - ϵ) / ϵ₁)
-        @ignore_derivatives alg.verbose > 1 && @printf(
-            "CTMRG: \titeration: %4d\t\terror: %.2e\t\tnorm: %.10e\t\tϵ: %.2e\t\tdϵ: %.2e\n",
-            iter,
-            err,
-            abs(new_norm),
-            ϵ,
-            dϵ
-        )
-
-        old_norm = new_norm
-        ϵ₁ = ϵ
-        iter += 1
+        # Update convergence criteria
+        normold = normnew
+        CSold = CSnew
+        TSold = TSnew
+        ϵold = iterinfo.ϵ
     end
 
-    #@ignore_derivatives @show iter, new_norm, err
-    @ignore_derivatives iter > alg.maxiter &&
-        alg.verbose > 0 &&
-        @warn "maxiter $(alg.maxiter) reached: error was $(err)"
-
-    return envs
+    return env
 end
 
-function left_move(
-    peps_above::InfinitePEPS{PType},
-    peps_below::InfinitePEPS{PType},
-    alg::CTMRG,
-    envs::CTMRGEnv,
-) where {PType}
-    corners::typeof(envs.corners) = copy(envs.corners)
-    edges::typeof(envs.edges) = copy(envs.edges)
-
-    above_projector_type = tensormaptype(spacetype(PType), 1, 3, storagetype(PType))
-    below_projector_type = tensormaptype(spacetype(PType), 3, 1, storagetype(PType))
+# One CTMRG iteration x′ = f(A, x)
+function ctmrg_iter(state, env::CTMRGEnv, alg::CTMRG)
     ϵ = 0.0
-    n0 = 1.0
-    n1 = 1.0
-    for col in 1:size(peps_above, 2)
-        cop = mod1(col + 1, size(peps_above, 2))
-        com = mod1(col - 1, size(peps_above, 2))
+    Pleft = Vector{Matrix{typeof(env.edges[1])}}(undef, 4)
```

Comment:
I wrote up the small `empty_projectors` function (naming should probably change) that creates empty projectors with the according types. I think it does make sense to have a separate function, since this pattern comes up in a few spots actually.

Reply:
`projector_type()` ?